tonilopezmr/day6/MemoryBank.kt

```diff
@@ -0,0 +1,104 @@
+package day6
+
+import kategory.combine
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+typealias P = Pair<Int, Int>
+
+class MemoryBank {
+
+  private fun chooseBank(input: List<Int>): Pair<Int, Int> =
+      input.foldRightIndexed(P(0, 0)) { idx, bank, (index, blocks) ->
+        if (bank >= blocks) P(idx, bank) else P(index, blocks)
+      }
+
+  private fun balance(input: List<Int>): List<Int> {
+    var (index, bank) = chooseBank(input)
+    val distribution = if (bank < input.size) bank else input.size - 1
+    val blocksForEach = bank / distribution
+    val list = input.toMutableList()
+    val displacement = blocksForEach * distribution
+    list[index] = bank - displacement
+    index++
+    for (i in 1..distribution) {
+      index %= input.size
+      list[index] = list[index] + blocksForEach
+      index++
+    }
+
+    return list
+  }
+
+  private fun infiniteLoopCircleInI(input: List<Int>): Int {
+    val history = mutableListOf<List<Int>>()
+    var banks = input
+    var count = 0
+    while (!history.contains(banks)) {
+      history.add(banks)
+      banks = balance(banks)
+      count++
+    }
+
+    return count
+  }
+
+  tailrec private fun infiniteLoopCircleIn(input: List<Int>, history: List<List<Int>> = emptyList(), count: Int = 0): Int =
+      if (history.contains(input)) count
+      else {
+        val newInput = balance(input)
+        infiniteLoopCircleIn(newInput, history.toMutableList().apply { add(input) }, count + 1)
+      }
```

Comment:
Recursion is the easiest way to go for this 😉

Reply:
but I don't know how to solve the history without mutability 🤔